src/repr/src/bytes.rs

```diff
@@ -0,0 +1,232 @@
+// Copyright Materialize, Inc. and contributors. All rights reserved.
+//
+// Use of this software is governed by the Business Source License
+// included in the LICENSE file.
+//
+// As of the Change Date specified in that file, in accordance with
+// the Business Source License, use of this software will be governed
+// by the Apache License, Version 2.0.
+
+use std::fmt::{self, Display};
+use std::str::FromStr;
+
+use serde::{Deserialize, Serialize};
+
+use mz_ore::cast::CastLossy;
+
+/// Struct for postgres compatible size units which is slightly different from
+/// [`bytesize::ByteSize`]. Instead of MiB or GiB and so on, it uses MB, GB for the sizes
+/// with 1024 multiplier.
+/// Valid units are B, kB, MB, GB, TB, PB with multiples of 1024
+/// where 1MB = 1024kB
+#[derive(Debug, Clone, PartialEq, Eq, Hash, PartialOrd, Ord, Serialize, Deserialize, Default)]
+pub struct Bytes(pub u64);
+
+impl Bytes {
+    pub const fn b(size: u64) -> Bytes {
+        Bytes(size)
+    }
+
+    pub const fn kb(size: u64) -> Bytes {
+        Bytes(size * BytesUnit::Kb.value())
+    }
+
+    pub const fn mb(size: u64) -> Bytes {
+        Bytes(size * BytesUnit::Mb.value())
+    }
+
+    pub const fn gb(size: u64) -> Bytes {
+        Bytes(size * BytesUnit::Gb.value())
+    }
+
+    pub const fn tb(size: u64) -> Bytes {
+        Bytes(size * BytesUnit::Tb.value())
+    }
+
+    pub const fn pb(size: u64) -> Bytes {
+        Bytes(size * BytesUnit::Pb.value())
+    }
+
+    pub fn as_u64(&self) -> u64 {
+        self.0
+    }
+    /// Copied logic from [`bytesize::ByteSize::to_string`]
+    /// with slight change to have no space between the number and the unit
+    /// to have the same behaviour as Postgres.
+    pub fn to_string(bytes: u64) -> String {
+        let unit = 1024;
+        let unit_base = 6.907755279; // ln 1000
+        let unit_prefix = ['k', 'M', 'G', 'T', 'P'];
+        let unit_suffix = 'B';
+
+        if bytes < unit {
+            format!("{}B", bytes)
+        } else {
+            let size: f64 = f64::cast_lossy(bytes);
+            let exp: usize = match usize::cast_lossy(size.ln() / unit_base) {
+                e if e == 0 => 1,
+                e => e,
+            };
+
+            format!(
+                "{:.1}{}{}",
+                size / f64::cast_lossy(unit.pow(exp.try_into().expect("usize into u32"))),
+                unit_prefix[exp - 1],
+                unit_suffix
+            )
+        }
+    }
+}
+
+impl Display for Bytes {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.pad(&Self::to_string(self.0))
+    }
+}
+
+impl FromStr for Bytes {
```

Comment:
Should `Bytes` roundtrip through it's string representation? If so can we add a `proptest` for that? Something like
```
proptest! {
  #[mz_ore::test]
  fn proptest_bytes_roundtrips_string(og: Bytes) {
    let s = og.to_string();
    let rnd = Bytes::from_str(&s).expect("roundtrip");
    prop_assert_eq!(og, rnd);
  }
}
```

Reply:
Yup, it should roundtrip! Let me add that

Reply:
I spoke too soon. It actually does not with fractional values. For ex `1025 bytes`, will still be shown as `1 kB`, which roundtrips back to `1024 bytes`. There are some lossy casts `u64` to `f64` and vice versa

Reply:
I added a String -> ByteSize - String roundtrip instead.

Reply:
Makes sense, thanks!